lib/login/index.js

```diff
@@ -66,6 +67,100 @@ if (config.login.twitch && config.login.twitch.enabled) {
 	}));
 }
 
+async function makeDiscordAPIRequest(guild, userID) {
+	const res = await fetch(`https://discord.com/api/v8/guilds/${guild.guildID}/members/${userID}`,
+		{
+			headers: {
+				Authorization: `Bot ${guild.guildBotToken}`
+			}
+		}
+	);
+	const data = await res.json();
+	if (res.status === 200) {
+		return [guild, false, data];
+	}
+
+	return [guild, true, data];
+}
+
+if (config.login.discord && config.login.discord.enabled) {
+	const DiscordStrategy = require('passport-discord').Strategy;
+
+	// The "identify" scope is required. Add it if not present.
+	let scope = config.login.discord.scope.split(' ');
+	if (scope.indexOf('identify') < 0) {
+		scope.push('identify');
+	}
+
+	// The "guilds" scope is required if allowedGuilds are used. Add it if not present.
+	if (scope.indexOf('guilds') < 0 && config.login.discord.allowedGuilds) {
+		scope.push('guilds');
+	}
+
+	scope = scope.join(' ');
+	passport.use(new DiscordStrategy({
+		clientID: config.login.discord.clientID,
+		clientSecret: config.login.discord.clientSecret,
+		callbackURL: `${protocol}://${config.baseURL}/login/auth/discord`,
+		scope
+	}, async (accessToken, refreshToken, profile, done) => {
+		if (config.login.discord.allowedUserIDs && config.login.discord.allowedUserIDs.includes(profile.id)) {
+			// Users that are on allowedUserIDs are allowed
+			profile.allowed = true;
+		} else if (config.login.discord.allowedGuilds) {
+			// Get guilds that are specified in the config and that user is in
+			const intersectingGuilds = config.login.discord.allowedGuilds.filter(allowedGuild => {
+				return profile.guilds.some(profileGuild => profileGuild.id === allowedGuild.guildID);
+			});
+
+			const guildRequests = [];
+
+			for (const intersectingGuild of intersectingGuilds) {
+				if (!intersectingGuild.allowedRoleIDs || intersectingGuild.allowedRoleIDs.length === 0) {
+					// If the user matches any guilds that only have member and not role requirements we do not need to make requests to the discord API
+					profile.allowed = true;
+				} else {
+					// Queue up all requests to the Discord API to improve speed
+					guildRequests.push(makeDiscordAPIRequest(intersectingGuild, profile.id));
+				}
+			}
```

Comment:
This ends up making unnecessary requests for guild info whether or not the user has a guild that doesn't have role IDs specified.

Reply:
It shouldn't as it just pushes the promise to the `guildRequests` array, the promises don't actually get run if `profile.allowed === true` in line 128, only if it's not allowed they get run in line 129